### PR TITLE
CSHARP-4772: Add support for List<T>.Exists and Array.Exists<T> in LINQ3.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            var organizationId = 1;
+
+            var queryable = collection.AsQueryable()
+                .Select(a =>
+                    a.ProfilesList != null &&
+                    a.ProfilesList.Any(p => p.OrganizationIdsList.Any(o => o == organizationId)));
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs:43]: A number of these scenarios are not supported by LINQ2.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            var stages = Translate(collection, queryable);
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                AssertStages(stages, "{ $match : { ProfilesList : { $ne : null }, 'ProfilesList.OrganizationIdsList' : { $elemMatch : { $eq : 1 } } } }");
+            }
+            else
+            {
+                AssertStages(stages, "{ $match : { ProfilesList : { $ne : null, $elemMatch : { OrganizationIdsList : 1 } } } }");
+            }
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs:131]: The MQL is different between LINQ2 and LINQ3 but I believe both of them to be correct.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs (hunk 1 above)
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs:43]: Should we check the exception somehow? If now, why do we even need LINQ2 provider here?

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            var organizationId = 1;
+
+            var queryable = collection.AsQueryable()
+                .Select(a =>
+                    a.ProfilesArray != null &&
+                    Array.Exists(a.ProfilesArray, p => Array.Exists(p.OrganizationIdsArray, o => o == organizationId)));
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs:70]: Same here. We should either check the exception if it's expected one, or remove LINQ2 code at all.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                new Account
+                {
+                    Id = 1,
+                    ProfilesArray = new Profile[] { new Profile { OrganizationIdsArray = new int[] { 1 } } },
+                    ProfilesList = new List<Profile> { new Profile { OrganizationIdsList = new List<int> { 1 } } }
+                },
+                new Account
+                {
+                    Id = 2,
+                    ProfilesArray = new Profile[] { new Profile { OrganizationIdsArray = new int[] { 2 } } },
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs:205]: Should we add tests with empty and null arrays and lists?

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs (hunk 4 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs:205]: I've added more test data.

**Review by sanych-sun — APPROVED** [review]
LGTM

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs (hunk 2 above)
Comment (JamesKovacs) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4772Tests.cs:133]: Agreed that they are equivalent.

**Review by JamesKovacs — APPROVED** [review]
LGTM
